data.rs

```diff
@@ -0,0 +1,170 @@
+use crate::PSP22Error;
+use ink::{
+    prelude::{vec, vec::Vec},
+    primitives::AccountId,
+    storage::Mapping,
+};
+
+pub enum PSP22Event {
+    Transfer(Option<AccountId>, Option<AccountId>, u128),
+    Approval(AccountId, AccountId, u128),
+}
+
+#[ink::storage_item]
+#[derive(Debug, Default)]
+pub struct PSP22Data {
+    total_supply: u128,
+    balances: Mapping<AccountId, u128>,
+    allowances: Mapping<(AccountId, AccountId), u128>,
+}
+
+impl PSP22Data {
+    pub fn new(supply: u128, creator: AccountId) -> PSP22Data {
+        let mut data = PSP22Data {
+            total_supply: supply,
+            balances: Default::default(),
+            allowances: Default::default(),
+        };
+        data.balances.insert(creator, &supply);
+        data
+    }
+
+    pub fn total_supply(&self) -> u128 {
+        self.total_supply
+    }
+
+    pub fn balance_of(&self, owner: AccountId) -> u128 {
+        self.balances.get(owner).unwrap_or_default()
+    }
+
+    pub fn allowance(&self, owner: AccountId, spender: AccountId) -> u128 {
+        self.allowances.get((owner, spender)).unwrap_or_default()
+    }
+
+    pub fn transfer(
+        &mut self,
+        caller: AccountId,
```

Comment:
```suggestion
        from: AccountId,
```

At this level of abstraction there's no notion of "caller" anymore.

Reply:
These internal methods of `PSP22Data` differ in signature form the ones defined by the `PSP22` trait. My idea was to make it explicit, that this argument is meant to always be `self.env().caller()` when someone is implementing the trait.